### PR TITLE
Price change notification

### LINKED ISSUE #443: Price change notification
Need to implement a mechanism which will notify Message Broker about price changes in Backoffcie

**AC:**
* Magento "Export API" sends a notification about price change via Message Bus
* Message contains required information about new price and current scope (website, customer group)
* Message should be sent only if the price was changed
* All prices changes should be honored
  * regular prices
  * special prices
  * tier price
  * custom option prices
  * the complex product option price (downloadable link, gift card amounts)
* Consumer accepts messages with price changes in the Message Broker 
  * No logic required, only accept message

**Tech details:**
* use mView based approach to react on price changes in the source tables
  * example of mView configuration <https://github.com/magento/saas-export/pull/117/files]
  * approach can be used after merging [https://github.com/magento/magento2/pull/30790]
  * mView should be used regardless on indexer modes (can be done in parallel in a separate task)
* The message format for price change:
  * product id list may be applied for mass price updates
  * topic name: "export.product.prices"
```java
[
  {
    meta: {
       type: "price_changed", 
       back_url: <ready-to-use URL to BackOffice to GetPrices API>,
    },
    data: [
        { 
         id: <product or option ids>,
         w:  <affected website>,
         cg: <affected customer group>,
       },
       ...
    ]
  
  },
   ...
]
{code}

* message format to send changes about tier prices updates (only with qty > 1, qty = 1 should be handled in previous case)
{code}
{
    meta: {
       type: "tier*price*changed", 
       // no back-url, data processed on Message Broker immediately 
    },
    data: [
      {
        id: <product id>,
        w: <affected website OR null for "all">,
        cg: <affected customer group OR null for "all">,
        qty: <qty>,
        price: <price>
      }, 
       ...
   ]

}  
{code}
* If tier price was deleted
{code}
{
    meta: {
       type: "price_deleted", 
       // no back-url, data processed on Message Broker immediately 
    },
    data: [
      {
        id: <product id>,
        w: <affected website OR null for "all">,
        cg: <affected customer group OR null for "all">,
        qty: <qty>, 
      }, 
       ...
   ]
}  
{code}
* If price  was deleted
{code}
{
    meta: {
       type: "price_deleted", 
       back_url: <ready-to-use URL to BackOffice to GetPrices API>,
    },
    data: [
      {
        id: [product ids],
        w: <affected website OR null for "all">,
        cg: <affected customer group OR null for "all">,
      }, 
       ...
   ]
}  
{code}
* If product was deleted. topic: "catalog.export.product.data"
{code}
{
    meta: {
       type: "product_deleted", 
       // no back-url, data processed on Message Broker immediately 
    },
    data: {ids: [affected ids]}
   >
}  
```

### LINKED ISSUE #469: [Prices] Gift cards price notification events
Price change notifications PR - https://github.com/magento/commerce-data-export/pull/20 .
Add the possibility to expose gift card prices change events.

### PR DESCRIPTION
**Pull request by aleksejsbaranovs10**
### Related Pull Requests
https://github.com/magento/magento2/pull/30790
https://github.com/magento/partners-magento2ee/pull/404

https://github.com/skovalenk/magento2/pull/1
https://github.com/skovalenk/partners-magento2ee/pull/1

### Fixed Issues (if relevant)

1. Fixes magento/catalog-storefront#443
2. Fixes magento/catalog-storefront#469

### Code Review Checklist (*)

- [ ] Story AC is completed
- [ ] proposed changes correspond to [Magento Technical Vision](https://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html)
- [ ] new or changed code is covered with web-api/integration tests (if applicable)
- [ ] no backward incompatible changes

### CODE REVIEW

**Comment by aleksejsbaranovs10** [comment]
@magento run all tests

**Comment by aleksejsbaranovs10** [comment]
@magento run Static Tests

**Comment by aleksejsbaranovs10** [comment]
@magento run all tests

File: app/code/Magento/CatalogDataExporter/Model/Provider/Product/ProductOptions/OptionValueUid.php [hunk 1]
```diff
@@ -1,38 +0,0 @@
-<?php
```
Comment (mslabko) [inline app/code/Magento/CatalogDataExporter/Model/Provider/Product/ProductOptions/OptionValueUid.php:1]: app/code/Magento/ProductVariantDataExporter/Test/Integration/AbstractProductVariantsTest.php
Please, delete usage of this class

**Comment by aleksejsbaranovs10** [comment]
@magento run all tests

**Comment by RuslanKostiv1** [comment]
@magento run all tests

**Comment by RuslanKostiv1** [comment]
```
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/commerce-data-export/pull/20/eec199f3c98e5c410bece73f0bf31f8a/Statics/console-error-logs.html
```

File: app/code/Magento/CatalogPriceDataExporter/Model/Event/ComplexProductEvent.php [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+     */
+    private function getEventData(array $indexData, array $actualData): array
+    {
+        $events = [];
+
+        foreach ($indexData as $data) {
+            $actualVariations = $actualData[$data['parent_id']] ?? [];
+            $eventType = \in_array($data['entity_id'], $actualVariations) ? self::EVENT_VARIATION_CHANGED
+                : self::EVENT_VARIATION_DELETED;
+            $websiteId = (string)$this->storeManager->getWebsite(WebsiteInterface::ADMIN_CODE)->getWebsiteId();
```
Comment (RuslanKostiv1) [inline app/code/Magento/CatalogPriceDataExporter/Model/Event/ComplexProductEvent.php:133]: will website id be always equal to admin website id?

File: app/code/Magento/CatalogPriceDataExporter/Model/Event/CustomOptionPriceEvent.php [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+use Magento\CatalogDataExporter\Model\Provider\Product\ProductOptions\CustomizableEnteredOptionValueUid;
+use Magento\CatalogPriceDataExporter\Model\EventKeyGenerator;
+use Magento\CatalogPriceDataExporter\Model\Query\CustomOptionPrice;
+use Magento\DataExporter\Exception\UnableRetrieveData;
+use Magento\Framework\App\ResourceConnection;
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\Store\Model\StoreManagerInterface;
+use Psr\Log\LoggerInterface;
+use Magento\Store\Api\Data\WebsiteInterface;
```
Comment (RuslanKostiv1) [inline app/code/Magento/CatalogPriceDataExporter/Model/Event/CustomOptionPriceEvent.php]: is not used, please remove

File: app/code/Magento/CatalogPriceDataExporter/Model/Event/CustomOptionTypePriceEvent.php [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+namespace Magento\CatalogPriceDataExporter\Model\Event;
+
+use Magento\CatalogDataExporter\Model\Provider\Product\ProductOptions\CustomizableSelectedOptionValueUid;
+use Magento\CatalogPriceDataExporter\Model\EventKeyGenerator;
+use Magento\CatalogPriceDataExporter\Model\Query\CustomOptionTypePrice;
+use Magento\DataExporter\Exception\UnableRetrieveData;
+use Magento\Framework\App\ResourceConnection;
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\Store\Api\Data\WebsiteInterface;
```
Comment (RuslanKostiv1) [inline app/code/Magento/CatalogPriceDataExporter/Model/Event/CustomOptionTypePriceEvent.php]: is not used, please remove

File: app/code/Magento/CatalogPriceDataExporter/Model/Event/ComplexProductEvent.php (hunk 2 above)
Comment (aleksejsbaranovs10) [reply app/code/Magento/CatalogPriceDataExporter/Model/Event/ComplexProductEvent.php:133]: yes, AFAIK not possible to change variation on website / store scope

File: app/code/Magento/CatalogDataExporter/Model/Provider/Product/ProductOptions/OptionValueUid.php (hunk 1 above)
Comment (aleksejsbaranovs10) [reply app/code/Magento/CatalogDataExporter/Model/Provider/Product/ProductOptions/OptionValueUid.php:1]: Done

File: app/code/Magento/CatalogPriceDataExporter/Model/Event/CustomOptionTypePriceEvent.php (hunk 4 above)
Comment (aleksejsbaranovs10) [reply app/code/Magento/CatalogPriceDataExporter/Model/Event/CustomOptionTypePriceEvent.php]: Done

File: app/code/Magento/CatalogPriceDataExporter/Model/Event/CustomOptionPriceEvent.php (hunk 3 above)
Comment (aleksejsbaranovs10) [reply app/code/Magento/CatalogPriceDataExporter/Model/Event/CustomOptionPriceEvent.php]: Done

**Comment by aleksejsbaranovs10** [comment]
@magento run all tests

**Comment by mslabko** [comment]
@magento import pr to magento-commerce/commerce-data-export

**Comment by magento-engcom-team** [comment]
@mslabko the pull request successfully imported.

**Pull request merged**